File: ChessLab/Analysis/GameResultPill.swift
import SwiftUI

/// Pastille de RÉSULTAT : elle RESSEMBLE au camp qui a gagné.
///
/// Née dans la bibliothèque, extraite ici parce que l'accueil montrait les
/// mêmes parties avec une pastille émeraude uniforme : deux écrans, deux
/// langages pour le même résultat. La palette est celle de l'échiquier —
/// ivoire pour les Blancs, ardoise pour les Noirs, moitié-moitié pour la
/// nulle — et non des couleurs sémantiques (vert/rouge), qui supposeraient
/// un camp « bon ».
///
/// Règles héritées de la bibliothèque, qui restent vraies partout :
/// - l'ivoire n'est PAS le blanc du texte, l'ardoise n'est PAS le fond ;
/// - l'ardoise porte un liseré clair, sans quoi elle disparaîtrait dans le
///   fond de la carte, lui-même sombre ;
/// - la couleur ne porte jamais seule : le texte reste lisible dans la
///   pastille et VoiceOver annonce « Victoire des Blancs » — sinon
///   l'information s'évanouit pour un daltonien comme pour un non-voyant.
struct GameResultPill: View {
    let raw: String?

    private enum ResultStyle {
        case white, black, draw, unknown

        static func of(_ raw: String?) -> ResultStyle {
            switch raw {
            case "1-0": .white
            case "0-1": .black
            case "1/2-1/2", "1/2": .draw
            default: .unknown
            }
        }

        var label: String {
            switch self {
            case .white: "1-0"
            case .black: "0-1"
            // « 1/2-1/2 » est deux fois plus large que les autres et
            // déformait l'alignement des lignes sans rien apprendre de plus.
            case .draw: "1/2"
            case .unknown: "?"
            }
        }
    }

    /// Ivoire de la pièce claire — pas le blanc du texte.
    private static let ivory = Color(red: 0.898, green: 0.882, blue: 0.839)
    /// Ardoise de la pièce sombre, plus claire que le fond de carte pour
    /// rester une forme et non un trou.
    private static let slate = Color(red: 0.110, green: 0.133, blue: 0.180)

    var body: some View {
        let style = ResultStyle.of(raw)
        Text(style.label)
            .font(.caption2.monospaced().weight(.bold))
            .foregroundStyle(foreground(style))
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            // Largeur commune : sans elle, « 1/2 » et « ? » rétrécissent la
            // pastille et les dates ne s'alignent plus d'une ligne à l'autre.
            .frame(minWidth: 34)
            .background { background(style) }
            .overlay {
                // Le liseré n'est utile qu'aux fonds sombres ou absents.
                if style != .white {
                    Capsule().strokeBorder(Color.white.opacity(0.22), lineWidth: 1)
                }
            }
            .clipShape(Capsule())
            .accessibilityLabel(Self.resultDescription(raw))
    }

    private func foreground(_ style: ResultStyle) -> Color {
        switch style {
        case .white: Theme.background
        case .black: Theme.textPrimary
        // Sur une pastille mi-ivoire mi-ardoise, ni le clair ni le sombre ne
        // tiennent partout : le texte reste blanc et l'ivoire est assez
        // atténué de son côté pour qu'il se lise.
        case .draw: Theme.textPrimary
        case .unknown: Theme.textSecondary
        }
    }

    @ViewBuilder
    private func background(_ style: ResultStyle) -> some View {
        switch style {
        case .white:
            Capsule().fill(Self.ivory)
        case .black:
            Capsule().fill(Self.slate)
        case .draw:
            // Partagée, littéralement : la moitié gauche revient aux Blancs.
            Capsule().fill(
                LinearGradient(
                    stops: [
                        .init(color: Self.ivory.opacity(0.55), location: 0),
                        .init(color: Self.ivory.opacity(0.55), location: 0.5),
                        .init(color: Self.slate, location: 0.5),
                        .init(color: Self.slate, location: 1),
                    ],
                    startPoint: .leading, endPoint: .trailing
                )
            )
        case .unknown:
            // Pas de résultat, pas de remplissage.
            Color.clear
        }
    }

    /// Ce que VoiceOver annonce : « 1-0 » ne se dit pas.
    static func resultDescription(_ raw: String?) -> String {
        switch raw {
        case "1-0": return LocalizationController.string("Victoire des Blancs")
        case "0-1": return LocalizationController.string("Victoire des Noirs")
        case "1/2-1/2", "1/2": return LocalizationController.string("Partie nulle")
        default: return LocalizationController.string("Résultat inconnu")
        }
    }
}

File: android/app/src/main/java/com/chesslab/ui/GameResultPill.kt
package com.chesslab.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * La pastille de résultat : ivoire pour les Blancs, ardoise pour les Noirs,
 * mi-l'un mi-l'autre pour la nulle. Pendant de `GameResultPill.swift` — le
 * MÊME langage partout où un résultat s'affiche, plutôt qu'un « 1/2-1/2 »
 * sur fond émeraude ici et un autre là.
 */
@Composable
fun GameResultPill(raw: String?, modifier: Modifier = Modifier) {
    val style = when (raw) {
        "1-0" -> Style.white
        "0-1" -> Style.black
        "1/2-1/2", "1/2" -> Style.draw
        else -> Style.unknown
    }
    val foreground = when (style) {
        Style.white -> Palette.background
        Style.black, Style.draw -> Palette.textPrimary
        Style.unknown -> Palette.textSecondary
    }
    val background: Brush = when (style) {
        Style.white -> Brush.linearGradient(listOf(ivory, ivory))
        Style.black -> Brush.linearGradient(listOf(slate, slate))
        // Sur une pastille mi-ivoire mi-ardoise, l'ivoire est atténué de son
        // côté pour que le texte clair s'y lise encore.
        Style.draw -> Brush.horizontalGradient(
            0f to ivory.copy(alpha = 0.55f), 0.5f to ivory.copy(alpha = 0.55f), 0.5f to slate, 1f to slate,
        )
        Style.unknown -> Brush.linearGradient(listOf(Color.Transparent, Color.Transparent))
    }
    Box(
        modifier
            // Largeur commune : sans elle, « 1/2 » et « ? » rétrécissent la
            // pastille et les dates ne s'alignent plus d'une ligne à l'autre.
            .widthIn(min = 34.dp)
            .clip(CircleShape)
            .background(background)
            // Le liseré n'est utile qu'aux fonds sombres ou absents.
            .then(if (style != Style.white) Modifier.border(1.dp, Color.White.copy(alpha = 0.22f), CircleShape) else Modifier)
            .padding(horizontal = 6.dp, vertical = 2.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            when (style) {
                Style.white -> "1-0"
                Style.black -> "0-1"
                // « 1/2-1/2 » est deux fois plus large et déformait l'alignement.
                Style.draw -> "1/2"
                Style.unknown -> "?"
            },
            fontSize = 10.sp, fontFamily = FontFamily.Monospace, fontWeight = FontWeight.Bold, color = foreground,
        )
    }
}

private enum class Style { white, black, draw, unknown }

/** Ivoire de la pièce claire — pas le blanc du texte. */
private val ivory = Color(0.898f, 0.882f, 0.839f)
/** Ardoise de la pièce sombre, plus claire que le fond de carte pour rester une forme et non un trou. */
private val slate = Color(0.110f, 0.133f, 0.180f)
